Source/Task/LocklessQueue.h

```diff
@@ -0,0 +1,740 @@
+    // the node pointer was preallocated.
+    //
+    void push_back(_In_ TData data, _In_ uint64_t address)
+    {
+        Address a;
+        a.value = address;
+        Node* n = to_node(a);
+        n->data = std::move(data);
+        m_activeList.push(n, a);
+    }
+    
+    //
+    // Pops TData off the head of the queue, returning
+    // true if successful or false if the queue is
+    // empty.
+    //
+    bool pop_front(_Out_ TData& data) noexcept
+    {
+        Address address;
+        Node* node = m_activeList.pop(address);
+        
+        if (node != nullptr)
+        {
+            data = std::move(node->data);
+            node->data = TData {};
+            m_heap.free(node, address);
+            return true;
+        }
+        
+        return false;
+    }
+
+    //
+    // Pops TData off the front of the queue, returning
+    // true if successful, and returns a node address
+    // instead of placing the address back on the free
+    // list. This address can later be used in a push_back
+    // call or a free_node call.  Failing to call either of these
+    // results in a leaked node. Returns false if the queue
+    // is empty.
+    //
+    bool pop_front(_Out_ TData& data, _Out_ uint64_t& address) noexcept
+    {
+        Address a;
+        Node* node = m_activeList.pop(a);
+        
+        if (node != nullptr)
+        {
+            data = std::move(node->data);
+            node->data = TData {};
+            address = a.value;
+            return true;
+        }
+        
+        return false;
+    }
+
+private:
+
+    /*
+     *
+     * Structure Definitions
+     *
+     */
+    
+    // Address - This represents the address of a node.
+    // Nodes live in a contiguous memory block, and there are multiple
+    // blocks. Address represents the position of the
+    // node and must be 64 bits.
+    struct Address
+    {
+        union
+        {
+            struct
+            {
+                uint32_t index;
+                uint16_t block;
+                uint16_t aba;
+            } s;
+            uint64_t value;
+        };
+        
+        inline bool operator == (_In_ const Address& other) const
+        {
+            return value == other.value;
+        }
+
+        inline bool operator != (_In_ const Address& other) const
+        {
+            return value != other.value;
+        }
+    };
+
+#if _MSVC_LANG >= 201703L
+    static_assert(std::atomic<Address>::is_always_lock_free, "LocklessQueue requires atomic<Address> to be lock free");
+#endif
+
+    // Node - each entry in a list is backed by a node,
+    // which contains a single 64 bit value representing
+    // the next node of a singly linked list and
+    // a payload of type TData. Nodes must be properly aligned
+    // in memory so std::atomic works consistently.  We do this
+    // by using an aligned allocator.
+    struct Node
+    {
+        std::atomic<Address> next;
+        TData data;
```

Comment:
should we only construct TData when data is actually pushed into the node and destroy it when it's popped? (conceptually like vector does it)

Reply:
There are definite advantages to that but I think not right now.  Your comment above about not calling dtors is important.  I've fixed that.  Demand constructing nodes would require that the block objects keep track of which nodes are allocated so they know which to destruct, or that block destruction be coordinated with list contents (which gets hard since lists can share blocks).

Reply:
That's fair